dm/syncer/validate_worker.go

```diff
@@ -350,45 +367,28 @@ func (vw *validateWorker) validateInsertAndUpdateRows(rows []*rowValidationJob,
 	return failedRows, nil
 }
 
-// a simplified version of https://github.com/pingcap/tidb-tools/blob/d9fdfa2f9040aab3fab7cd11774a82226f467fe7/sync_diff_inspector/utils/utils.go#L487-L606
-func (vw *validateWorker) compareData(sourceData, targetData []*sql.NullString, columns []*model.ColumnInfo) (bool, error) {
-	for i, column := range columns {
-		data1, data2 := sourceData[i], targetData[i]
-		if data1.Valid != data2.Valid {
-			return false, nil
-		}
-		str1, str2 := data1.String, data2.String
-		if str1 == str2 {
-			continue
-		} else if column.FieldType.Tp == tidbmysql.TypeFloat || column.FieldType.Tp == tidbmysql.TypeDouble {
-			// source and target data have different precision?
-			num1, err1 := strconv.ParseFloat(str1, 64)
-			num2, err2 := strconv.ParseFloat(str2, 64)
-			if err1 != nil || err2 != nil {
-				// should not happen
-				return false, errors.Errorf("convert %s, %s to float failed, err1: %v, err2: %v", str1, str2, err1, err2)
-			}
-			if math.Abs(num1-num2) <= 1e-6 {
-				continue
-			}
-		}
-		return false, nil
-	}
-
-	return true, nil
-}
-
 func (vw *validateWorker) getTargetRows(cond *Cond) (map[string][]*sql.NullString, error) {
-	tctx := tcontext.NewContext(vw.ctx, vw.L)
+	ctx, cancelFunc := context.WithTimeout(vw.ctx, queryTimeout)
+	defer cancelFunc()
+	tctx := tcontext.NewContext(ctx, vw.L)
 	columnNames := make([]string, 0, len(cond.Columns))
 	for _, col := range cond.Columns {
 		columnNames = append(columnNames, dbutil.ColumnName(col.Name.O))
 	}
 	columns := strings.Join(columnNames, ", ")
 	rowsQuery := fmt.Sprintf("SELECT /*!40001 SQL_NO_CACHE */ %s FROM %s WHERE %s",
 		columns, cond.TargetTbl, cond.GetWhere())
-	rows, err := vw.conn.QuerySQL(tctx, rowsQuery, cond.GetArgs()...)
+	// query using sql.DB directly, BaseConn is more than what we need
+	rows, err := vw.db.QueryContext(tctx, rowsQuery, cond.GetArgs()...)
```

Comment:
Do we need some extra tricks to handle retryable errors (e.g. connection error) here?

Reply:
no need to retry inside, worker will retry it anyway. see defer in `validateTableChange`